core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -628,6 +628,9 @@ private[spark] class TaskSetManager(
     // here "result.value()" just returns the value and won't block other threads.
     sched.dagScheduler.taskEnded(
       tasks(index), Success, result.value(), result.accumUpdates, info, result.metrics)
+    // kill other attempts of this task when speculation is enabled.
+    taskAttempts(index).filter(_.taskId != tid).foreach(e=>sched.killSpeculatedTask(e.taskId))
```

Comment:
It'd be slightly better to rewrite this to 

```
for (e <- taskAttempts(index) if e.taskId != tid) 
  sched.killSpeculatedTast..
```

The Scala compiler would translate the if-guard to a `withFilter` call, which is more efficient in that it won't create a new collection (which filter does).